x/p2p/router.go

```diff
@@ -0,0 +1,214 @@
+	}
+
+	client := &Client{
+		handlerID: handlerID,
+		sender:    r.sender,
+		router:    r,
+	}
+
+	responder := responder{
+		handler: handler,
+		sender:  r.sender,
+	}
+
+	r.handlers[handlerID] = responder
+
+	return client, nil
+}
+
+func (r *Router) AppRequest(ctx context.Context, nodeID ids.NodeID, requestID uint32, deadline time.Time, request []byte) error {
+	r.lock.RLock()
+	handlerID, parsedMsg, handler, ok := r.parse(request)
+	r.lock.RUnlock()
+	if !ok {
+		return fmt.Errorf("failed to process app request message for app protocol %d: %w", handlerID, ErrUnregisteredHandler)
+	}
+
+	return handler.AppRequest(ctx, nodeID, requestID, deadline, parsedMsg)
+}
+
+func (r *Router) AppRequestFailed(_ context.Context, nodeID ids.NodeID, requestID uint32) error {
+	r.lock.RLock()
+	callback, ok := r.clearAppRequest(requestID)
+	r.lock.RUnlock()
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	callback(nodeID, nil, ErrAppRequestFailed)
+	return nil
+}
+
+func (r *Router) AppResponse(_ context.Context, nodeID ids.NodeID, requestID uint32, response []byte) error {
+	r.lock.RLock()
+	callback, ok := r.clearAppRequest(requestID)
+	r.lock.RUnlock()
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	callback(nodeID, response, nil)
+	return nil
+}
+
+func (r *Router) AppGossip(ctx context.Context, nodeID ids.NodeID, gossip []byte) error {
+	r.lock.RLock()
+	handlerID, parsedMsg, handler, ok := r.parse(gossip)
+	r.lock.RUnlock()
+	if !ok {
+		return fmt.Errorf("failed to process gossip message for app protocol %d: %w", handlerID, ErrUnregisteredHandler)
+	}
+
+	return handler.AppGossip(ctx, nodeID, parsedMsg)
+}
+
+func (r *Router) CrossChainAppRequest(
+	ctx context.Context,
+	chainID ids.ID,
+	requestID uint32,
+	deadline time.Time,
+	msg []byte,
+) error {
+	r.lock.RLock()
+	handlerID, parsedMsg, handler, ok := r.parse(msg)
+	r.lock.RUnlock()
+
+	if !ok {
+		return fmt.Errorf("failed to process cross chain app request message for app protocol %d: %w", handlerID, ErrUnregisteredHandler)
+	}
+
+	return handler.CrossChainAppRequest(ctx, chainID, requestID, deadline, parsedMsg)
+}
+
+func (r *Router) CrossChainAppRequestFailed(_ context.Context, chainID ids.ID, requestID uint32) error {
+	r.lock.RLock()
+	callback, ok := r.clearCrossChainAppRequest(requestID)
+	r.lock.RUnlock()
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	callback(chainID, nil, ErrAppRequestFailed)
+	return nil
+}
+
+func (r *Router) CrossChainAppResponse(_ context.Context, chainID ids.ID, requestID uint32, response []byte) error {
+	r.lock.RLock()
+	callback, ok := r.clearCrossChainAppRequest(requestID)
+	r.lock.RUnlock()
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	callback(chainID, response, nil)
+	return nil
+}
+
+// Parse parses a gossip or request message and maps it to a corresponding
+// handler if present.
+func (r *Router) parse(msg []byte) (byte, []byte, responder, bool) {
```

Comment:
I feel like we should move the lock grabbing into `parse`. Should be able to reduce the scope of the lock and reduce the number of places we interact with it as well afaict